### PR TITLE
New patch release SLEPc 3.18.3

### PR DESCRIPTION
**Pull request by joseeroman**

### CODE REVIEW

**Review by balay — DISMISSED** [review]

**Review by adamjstewart — APPROVED** [review]

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Actually, I believe this is missing:
```python
depends_on("py-cython@0.24:", type="build")
```

**Comment by joseeroman** [comment]
> Actually, I believe this is missing:
> 
> ```python
> depends_on("py-cython@0.24:", type="build")
> ```

In the previous patch release we added this
```
depends_on("py-cython@0.24:", type="build", when="@main")
```
Do you mean I should remove the `when` argument?

**Comment by adamjstewart** [comment]
Either remove `when` if the dependency was always necessary, or change it to `@X.Y.Z:` where X.Y.Z is the first version where cython was required to install.

**Comment by balay** [comment]
@joseeroman  similar issue [and change] at https://github.com/spack/spack/pull/36406

**Comment by joseeroman** [comment]
@balay this happens also in petsc4py. From which version is cython 0.24 needed?

**Comment by joseeroman** [comment]
Ah, thanks. Will change.

**Review by tldahlgren — COMMENTED** [review]
Confirmed the new sha256.

**Review by adamjstewart — DISMISSED** [review]

**Comment by balay** [comment]
@joseeroman here are the changes from https://github.com/spack/spack/pull/36406 that work with py-slepc4py

```diff
diff --git a/var/spack/repos/builtin/packages/py-slepc4py/ldshared-dev.patch b/var/spack/repos/builtin/packages/py-slepc4py/ldshared-dev.patch
index fbe430af1b..be5b2569aa 100644
--- a/var/spack/repos/builtin/packages/py-slepc4py/ldshared-dev.patch
+++ b/var/spack/repos/builtin/packages/py-slepc4py/ldshared-dev.patch
@@ -1,7 +1,7 @@
 diff --git a/src/binding/slepc4py/conf/baseconf.py b/src/binding/slepc4py/conf/baseconf.py
 index b0707a65a..44ba8f194 100644
---- a/src/binding/slepc4py/conf/baseconf.py
-+++ b/src/binding/slepc4py/conf/baseconf.py
+--- a/src/binding/slepc4py/conf/confpetsc.py
++++ b/src/binding/slepc4py/conf/confpetsc.py
 @@ -213,7 +213,7 @@ class PetscConfig:
          ldshared = getenv('LDSHARED', ldshared)
          ldflags = getenv('LDFLAGS', cflags + ' ' + (ldflags or ''))
diff --git a/var/spack/repos/builtin/packages/py-slepc4py/package.py b/var/spack/repos/builtin/packages/py-slepc4py/package.py
index 293242612f..3a7924da48 100644
--- a/var/spack/repos/builtin/packages/py-slepc4py/package.py
+++ b/var/spack/repos/builtin/packages/py-slepc4py/package.py
@@ -37,8 +37,8 @@ class PySlepc4py(PythonPackage):
     patch("ldshared.patch", when="@:99")
     patch("ldshared-dev.patch", when="@main")
 
-    depends_on("py-cython@0.29.32:", when="@3.19: ^python@3.11:", type="build")
-    depends_on("py-cython@0.24:", when="@3.19:", type="build")
+    depends_on("py-cython@0.29.32:", when="^python@3.11:", type="build")
+    depends_on("py-cython@0.24:", type="build")
     depends_on("py-setuptools", type="build")
 
     depends_on("py-petsc4py", type=("build", "run"))
@@ -67,3 +67,8 @@ def build_directory(self):
             return os.path.join(self.stage.source_path, "src", "binding", "slepc4py")
         else:
             return self.stage.source_path
+
+    @run_before("install")
+    def cythonize(self):
+        with working_dir(self.build_directory):
+            python(join_path("conf", "cythonize.py"))
```

Additionally:
```diff
diff --git a/var/spack/repos/builtin/packages/py-slepc4py/package.py b/var/spack/repos/builtin/packages/py-slepc4py/package.py
index a23002aac2..9c2ce75f8c 100644
--- a/var/spack/repos/builtin/packages/py-slepc4py/package.py
+++ b/var/spack/repos/builtin/packages/py-slepc4py/package.py
@@ -40,6 +40,7 @@ class PySlepc4py(PythonPackage):
     depends_on("py-setuptools", type="build")
 
     depends_on("py-petsc4py", type=("build", "run"))
+    depends_on("py-petsc4py@main", when="@main", type=("build", "run"))
     depends_on("py-petsc4py@3.18.0:3.18", when="@3.18.0:3.18", type=("build", "run"))
     depends_on("py-petsc4py@3.17.0:3.17", when="@3.17.0:3.17", type=("build", "run"))
     depends_on("py-petsc4py@3.16.0:3.16", when="@3.16.0:3.16", type=("build", "run"))
@@ -49,6 +50,7 @@ class PySlepc4py(PythonPackage):
     depends_on("py-petsc4py@3.11.0:3.11", when="@3.11.0:3.11", type=("build", "run"))
 
     depends_on("slepc")
+    depends_on("slepc@main", when="@main")
     depends_on("slepc@3.18.0:3.18", when="@3.18.0:3.18")
     depends_on("slepc@3.17.0:3.17", when="@3.17.0:3.17")
     depends_on("slepc@3.16.0:3.16", when="@3.16.0:3.16")
```

**Comment by joseeroman** [comment]
Pushed. Thanks.

**Review by balay — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-slepc4py/package.py [hunk 1]
```diff
@@ -36,10 +37,12 @@ class PySlepc4py(PythonPackage):
     patch("ldshared.patch", when="@:99")
     patch("ldshared-dev.patch", when="@main")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-slepc4py/package.py:38]: Out of curiosity, where are these patches needed, and is there any way to solve the issue in slepc/petsc main?

**Review by adamjstewart — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-slepc4py/package.py (hunk 1 above)
Comment (balay) [reply var/spack/repos/builtin/packages/py-slepc4py/package.py:38]: I think the code attempts to handle certain corner cases [perhaps with intel compilers?]. And this breaks with spack - as it changes compilers underneath - hence this fix to skip that code during spack build.

@dalcinl I don' think this change will affect other usages - so perhaps ok to add to petsc/slepc mainline?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-slepc4py/package.py:38]: I feel bad hardcoding things in petsc/slepc just for Spack (although maybe Spack is one of the more popular ways of installing things). I do frequently see patches specific to Homebrew, but those also break things for Spack. Maybe a more general solution that checks if the compiler is a symlink or something.

Comment (dalcinl) [reply var/spack/repos/builtin/packages/py-slepc4py/package.py:38]: I don't have strong opinions about this. Most of these hideous hacks were done just to workaround limitations of Python's distutils.

Comment (balay) [reply var/spack/repos/builtin/packages/py-slepc4py/package.py:38]: With the change in file names [that was reflected in one of the changes in this PR] - I think we'll need another patch file added to spack for 3.19. So if we can add this to petsc/slepc mainline - then we can avoid adding additional copy of this patch to spack.

I'll create a MR for petsc

**Pull request merged**